scripts/gen-index.cts

```diff
@@ -0,0 +1,148 @@
+import { readdir, writeFile } from 'node:fs/promises';
+import { join, parse, relative, resolve, type ParsedPath } from 'node:path';
+import { sep as posixSep } from 'node:path/posix';
+import { sep as winSep } from 'node:path/win32';
+import ts from 'typescript';
+
+const PRINTER = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
+const WRITE_MODE = process.argv.includes('-w');
+const PACKAGE_NAME = process.argv.at(WRITE_MODE ? 3 : 2);
+
+function isExported(node: ts.Declaration): boolean {
+	return (ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export) > 0;
+}
+
+function isType(node: ts.Node): boolean {
+	return [ts.SyntaxKind.InterfaceDeclaration, ts.SyntaxKind.TypeAliasDeclaration].includes(node.kind);
+}
+
+interface ModuleExportNodes {
+	normal: ts.Declaration[];
+	types: ts.Declaration[];
+}
+
+interface ModuleExportInclusions {
+	useModule: boolean;
+	useNormal: boolean;
+	useTypes: boolean;
+}
+
+class ModuleFile {
+	public readonly exports: ModuleExportNodes;
+	public readonly isPrivate: boolean;
+	public readonly exportInclusions: ModuleExportInclusions;
+	public readonly path: ParsedPath;
+	public readonly sourceFile: ts.SourceFile;
+
+	public constructor(sourceFile: ts.SourceFile) {
+		this.sourceFile = sourceFile;
+		this.path = parse(sourceFile.fileName);
+		this.isPrivate = this.path.name.startsWith('_');
+		this.exports = this.parseExports();
+		this.exportInclusions = this.deriveExportInclusions();
+	}
+
+	public generateExportSpecifiers(which: keyof ModuleExportNodes) {
+		return [...new Set(this.exports[which].map((node) => ts.getNameOfDeclaration(node)!.getText(this.sourceFile)!))].map((name) =>
+			ts.factory.createExportSpecifier(false, undefined, name)
+		);
+	}
+
+	/**
+	 * Rules:
+	 * 1. If the module begins with `_`, ignore it
+	 *    a) unless the module contains types, in which case those exclusively should be included
+	 * 2. If the module exclusively contains types, use `export type *`, otherwise use `export *`
+	 **/
+	public toExportDeclaration(packageDir: string): ts.ExportDeclaration | undefined {
+		const { useModule, useNormal, useTypes } = this.exportInclusions;
+		if (!useModule) return;
+
+		// TODO: make this more efficient
```

Comment:
Resolve in this PR, or delete the TODO. Your call. Adding a random TODO is poor management of technical debt. Taking up tech. debt is something we only do if there is a good reason to do so.